### PR TITLE
Add Redis output

### PR DESCRIPTION
**Pull request by pandyamarut**
Signed-off-by: pandyamarut [pandyamarut@gmail.com](mailto:pandyamarut@gmail.com) ([cloudDefense.ai](https://www.clouddefense.ai/))

**What type of PR is this?**

 /kind feature

**Any specific area of the project related to this PR?**

/area outputs

**What this PR does / why we need it**:
This introduces Redis output

### CODE REVIEW

**Review by anshubansal2000 — DISMISSED** [review]

File: README.md [hunk 1]
```diff
@@ -1033,6 +1040,11 @@ order is
 - **TIMESCALEDB_DATABASE**: TimescaleDB database used
 - **TIMESCALEDB_HYPERTABLENAME**: Hypertable to store data events (default: falco_events) See TimescaleDB setup for more info
 - **TIMESCALEDB_MINIMUMPRIORITY**: minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+- **REDIS_HOST**: Redis host, if not empty, Redis output is enabled
+- **REDIS_PORT**: Redis port (default: 6379)
```
Comment (Issif) [inline README.md:1044]: See my previous comment, we may merge them

File: README.md [hunk 2]
```diff
@@ -561,6 +561,13 @@ timescaledb:
   # database: "" # TimescaleDB database used
   # hypertablename: "falco_events" # Hypertable to store data events (default: falco_events) See TimescaleDB setup for more info
   # minimumpriority: "" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+redis:
+  # host: "" # Redis host, if not empty, Redis output is enabled
+  # port: "6379" # TimescaleDB port (default: 5432)
```
Comment (Issif) [inline README.md:566]:
```suggestion
  # port: "6379" # Redis port (default: 6379)
```

File: outputs/redis.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+func NewRedisClient(config *types.Configuration, stats *types.Statistics, promStats *types.PromStatistics,
+	statsdClient, dogstatsdClient *statsd.Client) (*Client, error) {
+
+	ctx := context.Background()
+
+	rClient := redis.NewClient(&redis.Options{
+		Addr:     config.Redis.Address + ":" + config.Redis.Port,
+		Password: config.Redis.Password,
```
Comment (Issif) [inline outputs/redis.go:21]: Is it working even if `config.Redis.Password` is empty?

File: outputs/redis.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+		DB:       config.Redis.Database,
+	})
+
+	// Ping the Redis server to check if it's running
+	pong, err := rClient.Ping(ctx).Result()
+	if err != nil {
+		log.Println(err)
```
Comment (Issif) [inline outputs/redis.go:28]: it doesn't follow the global log pattern

File: outputs/redis.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	fmt.Println(pong, "Redis Server Running")
```
Comment (Issif) [inline outputs/redis.go:30]: it doesn't follow the global log pattern

File: outputs/redis.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+func (c *Client) RedisPost(falcopayload types.FalcoPayload) {
+	c.Stats.Redis.Add(Total, 1)
+
+	ctx := context.Background()
+
+	redisPayload, _ := json.Marshal(falcopayload)
+
+	_, err := c.RedisClient.RPush(ctx, "Falco_Logs", redisPayload).Result()
```
Comment (Issif) [inline outputs/redis.go:50]: if we use a context, it should the same the client's

File: outputs/redis.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+		DogstatsdClient: dogstatsdClient,
+	}, nil
+}
+
+func (c *Client) RedisPost(falcopayload types.FalcoPayload) {
+	c.Stats.Redis.Add(Total, 1)
+
+	ctx := context.Background()
+
+	redisPayload, _ := json.Marshal(falcopayload)
```
Comment (Issif) [inline outputs/redis.go:48]: I'm not a Redis expert, but I understand we store the json as a simple string. Can we think about allowing other storage formats, like direct a `JSON` with https://redis.io/docs/stack/json/ or `hash(map)` https://redis.com/redis-best-practices/data-storage-patterns/object-hash-storage/.
The choice of the format could be set by the user too.

File: README.md [hunk 8]
```diff
@@ -561,6 +561,13 @@ timescaledb:
...
+  # password: "" # Password to authenticate with Redis (default: "")
+  # database: "" # Redis database number (default: 0)
+  # minimumpriority: "" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+
```
Comment (Issif) [inline README.md:570]: what about to allow to set a  `ttl` for keys?

**Review by Issif — CHANGES_REQUESTED** [review]

File: outputs/redis.go (hunk 3 above)
Comment (pandyamarut) [reply outputs/redis.go:21]: Yes @Issif, It does.

File: README.md [hunk 9]
```diff
@@ -561,6 +561,13 @@ timescaledb:
   # database: "" # TimescaleDB database used
   # hypertablename: "falco_events" # Hypertable to store data events (default: falco_events) See TimescaleDB setup for more info
   # minimumpriority: "" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+redis:
+  # host: "" # Redis host, if not empty, Redis output is enabled
```
[the remark this answers is not in the document]
Comment (pandyamarut) [reply README.md:565]: @Issif Right , I will address this.

File: README.md (hunk 8 above)
Comment (pandyamarut) [reply README.md:570]: @Issif , I am not sure if that can be done through redis config.

File: outputs/redis.go (hunk 7 above)
Comment (pandyamarut) [reply outputs/redis.go:48]: @Issif yes, We can. Thanks for the review, I will add the option for this well. I can set this in a config and set default as a list (which is currently there.)

File: outputs/redis.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+	rClient := redis.NewClient(&redis.Options{
+		Addr:     config.Redis.Address,
+		Password: config.Redis.Password,
+		DB:       config.Redis.Database,
+	})
+	// Ping the Redis server to check if it's running
+	pong, err := rClient.Ping(context.Background()).Result()
+	if err != nil {
+		log.Printf("[ERROR] : Redis misconfiguration, cannot connect to the server %v\n", err)
```
Comment (Issif) [inline outputs/redis.go:33]:
```suggestion
		log.Printf("[ERROR] : Redis - Misconfiguration, cannot connect to the server %v\n", err)
```

File: outputs/redis.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+		StatsdClient:    statsdClient,
+		DogstatsdClient: dogstatsdClient,
+	}, nil
+}
+
+func (c *Client) RedisPost(falcopayload types.FalcoPayload) {
+	c.Stats.Redis.Add(Total, 1)
+	redisPayload, _ := json.Marshal(falcopayload)
+	if strings.ToLower(c.Config.Redis.StorageType) == "hashmap" {
+		_, err := c.RedisClient.HSet(context.Background(), "Falco_logs", falcopayload.UUID, redisPayload).Result()
```
Comment (Issif) [inline outputs/redis.go:52]: Why is `Falco_logs` not a configurable value?

**Review by Issif — CHANGES_REQUESTED** [review]
You linter changed a lot of things in the README, can you revert them please, it creates a flood in that PR and a drift with `config_example.yaml` which is supposed to be updated too with the new  Redis section. Thanks

File: README.md [hunk 12]
```diff
@@ -561,6 +561,13 @@ timescaledb:
   # database: "" # TimescaleDB database used
   # hypertablename: "falco_events" # Hypertable to store data events (default: falco_events) See TimescaleDB setup for more info
   # minimumpriority: "" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+redis:
+  # address: "" # Redis address, if not empty, Redis output is enabled
+  # password: "" # Password to authenticate with Redis (default: "")
+  # database: "" # Redis database number (default: 0)
+  # storagetype: "" # Redis storage type hashmap or list (default)
```
Comment (Issif) [inline README.md:568]:
```suggestion
  # storagetype: "" # Redis storage type: hashmap or list (default: list)
```

File: README.md [hunk 13]
```diff
@@ -1033,6 +1040,12 @@ order is
 - **TIMESCALEDB_DATABASE**: TimescaleDB database used
 - **TIMESCALEDB_HYPERTABLENAME**: Hypertable to store data events (default: falco_events) See TimescaleDB setup for more info
 - **TIMESCALEDB_MINIMUMPRIORITY**: minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+- **REDIS_ADDRESS**: Redis host, if not empty, Redis output is enabled (example: localhost:6379)
+- **REDIS_PASSWORD**: Password to authenticate with Redis (default: "")
+- **REDIS_DATABASE**: Redis database number (default: 0)
+- **REDIS_STORAGE**: Redis storage type for ex: Hashmap, ,list (default)
```
Comment (Issif) [inline README.md:1046]:
```suggestion
- **REDIS_STORAGE**: Redis storage type: hashmap or list (default: list)
```

File: config.go [hunk 14]
```diff
@@ -402,6 +402,15 @@ func getConfig() *types.Configuration {
 	v.SetDefault("TimescaleDB.Hypertable", "falcosidekick_events")
 	v.SetDefault("TimescaleDB.MinimumPriority", "")
 
+	v.SetDefault("Redis.Address", "")
+	v.SetDefault("Redis.Password", "")
+	v.SetDefault("Redis.Database", 0)
+	v.SetDefault("Redis.StorageType", "")
```
Comment (Issif) [inline config.go:408]:
```suggestion
	v.SetDefault("Redis.StorageType", "list")
```

File: outputs/redis.go [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+		Addr:     config.Redis.Address,
+		Password: config.Redis.Password,
+		DB:       config.Redis.Database,
+	})
+	// Ping the Redis server to check if it's running
+	pong, err := rClient.Ping(context.Background()).Result()
+	if err != nil {
+		log.Printf("[ERROR] : Redis - Misconfiguration, cannot connect to the server %v\n", err)
+	}
+	log.Printf("[INFO] : Redis - Connected to redis server: %v", pong)
```
Comment (Issif) [inline outputs/redis.go:35]:
```suggestion
	log.Printf("[INFO]  : Redis - Connected to redis server: %v\n", pong)
```

**Review by Issif — APPROVED** [review]

**Review by Issif — APPROVED** [review]

**Review by Issif — APPROVED** [review]

**Pull request merged**